### ISSUE TITLE
simplify and document instructions for weave on a Mac

### ISSUE THREAD

**Issue by abuehrle** [weaveworks/weave#1285]
@squaremo pointed out that boot2docker uses tcp instead of unix sockets to run a proxy. Therefor, the user must set their DOCKER_HOST manually and also needs to run 'weave launch-proxy' after launching weave.

**Comment by rade** [comment]
We have instructions at http://docs.weave.works/weave/master/head/boot2docker.html, ...which aren't linked to from anywhere!

Furthermore, they diverge from the remainder of the docs in that they use `./weave` instead of `weave` and `eval "$(./weave env)"` instead of `eval $(weave env)`.

We MUST come up with a better way for launching weave on a b2d install.

**Comment by rade** [comment]
Now @abuerhrle is telling me that docker-machine is the way to go, since it "wraps up boot2docker and installs it for you on VirtualBox.". Is that what we recommend people use? Seems like a lot of machinery (pun intended). Is there really no shorter route?

And if machine really is the way to go, how is one meant to start the proxy? http://weave.works/guides/weave-and-docker-platform/chapter1/machine.html predates the proxy, so makes no mention of that.

In conclusion, right now we have zero information in our docs on how to make weave 1.0 (let alone 1.1) work on a Mac, which is what most of our target audience is using.

**Comment by bboreham** [comment]
Docker machine is indeed intended as the replacement for the boot2docker CLI.  I filed an [issue](https://github.com/boot2docker/boot2docker/issues/988) that this is not made very clear.

Deprecation notice [here](https://github.com/boot2docker/boot2docker-cli#pending-deprecation).  Note a hint [here](https://github.com/boot2docker/boot2docker/pull/1013/files) that they are also replacing the distro.

**Comment by rade** [comment]
Right. http://docs.docker.com/mac/step_one/ uses the Docker Toolbox, which includes machine.

**Comment by abuehrle** [comment]
@errordeveloper  had referenced this also.   I will try it out today.

**Comment by rade** [comment]
So where we want to get to here is the following sequence:

1. install docker as per docker's instructions, i.e. using the Docker Toolbox
2. download weave as per our normal instructions
3. perform some extra step, if necessary
4. launch weave as per our normal instructions

**Comment by abuehrle** [comment]
downloaded the dockertoolbox, and installed docker, docker-machine, boot2docker using that newish tool. I created an instance on Virtualbox using docker-machine, called weave-1, and then set up my env by running with the following: 
eval "$(docker-machine env weave-1)"
which sets the following: 

export DOCKER_TLS_VERIFY="1"
export DOCKER_HOST="tcp://192.168.99.101:2376"  <--- this is wrong, correct?-->
export DOCKER_CERT_PATH="/Users/anitab/.docker/machine/machines/weave-1"
export DOCKER_MACHINE_NAME="weave-1"

then ran: ./weave launch

DNS starts, but proxy does not, saying: 

When launching the proxy via TLS, -H and/or TLS options are required.

then, tried: 

./weave launch-proxy -H tcp://192.168.99.101:2376 $DOCKER_CERT_PATH $DOCKER_TLS_VERIFY

The weaveproxy container has died. Consult the logs with 'docker logs weaveproxy' for further details.

which shows: 

docker logs weaveproxy
INFO: 2015/08/11 20:37:54.413209 weave proxy git-94af02fbf41c
INFO: 2015/08/11 20:37:54.413349 Command line arguments: -H tcp://192.168.99.101:2376 /Users/anitab/.docker/machine/machines/weave-1 1
FATA: 2015/08/11 20:37:54.433962 Could not start proxy: Could not find the weavewait volume

so then, I tried:  

 eval "$(./weave env)"

and it returns: 

weaveproxy container is not present. Have you launched it?

What do I set the DOCKER_HOST proxy to?

I also tried to set the docker host to:  tcp://127.0.0.1:9999 

but that didn't work either.

**Comment by rade** [comment]
The `-H` needs to be set to where you want to the proxy to listen. Nope, no idea what that should be.

You can't just pass the $DOCKER_CERT_PATH and $DOCKER_TLS_VERIFY; you need to pass options along the lines shown in http://docs.weave.works/weave/master/head/boot2docker.html. And these options need to reference files on your docker host, *not* your client machine (which is where DOCKER_CERT_PATH is located). Nope, no idea what they should be either.

This is impossible for mere mortals to configure.

**Comment by abuehrle** [comment]
tlsargs=$(docker-machine ssh weave-1 \
>   "cat /proc/\$(pgrep /usr/local/bin/docker)/cmdline | tr '\0' '\n' | grep ^--tls | tr '\n' ' '")

and then, 

./weave launch-proxy $tlsargs
The weaveproxy container has died. Consult the logs with 'docker logs weaveproxy' for further details.

doesn't seem to work either: 

docker logs weaveproxy
INFO: 2015/08/11 21:26:36.897225 weave proxy git-94af02fbf41c
INFO: 2015/08/11 21:26:36.897625 Command line arguments: --tlsverify --tlscacert /home/weave/tls/ca.pem --tlscert /home/weave/tls/cert.pem --tlskey /home/weave/tls/key.pem -H tcp://0.0.0.0:12375
FATA: 2015/08/11 21:26:37.076987 Could not start proxy: Could not find the weavewait volume

**Comment by rade** [comment]
That looks pretty close actually. The weavewait thing is weird. What does `docker inspect weaveproxy` return?

**Comment by rade** [comment]
(and where on earth did you find that magic tlsargs line?!?)

**Comment by abuehrle** [comment]
That was from @errordevelopers or were they fintan’s docs.  

Here: 

http://weave.works/guides/weave-and-docker-platform/chapter2/machine-with-weave-proxy.html <http://weave.works/guides/weave-and-docker-platform/chapter2/machine-with-weave-proxy.html>

**Comment by abuehrle** [comment]
I think I see the address: 

 docker inspect weaveproxy
[
{
    "Id": "c269e54807befc16b6e0e53fe280cf6a4ca5f8eb0a0df9e0a69f5af0598fb0c8",
    "Created": "2015-08-11T21:34:22.248746543Z",
    "Path": "/home/weave/weaveproxy",
    "Args": [
        "--tls",
        "--tlscacert",
        "/home/weave/tls/ca.pem",
        "--tlscert",
        "/home/weave/tls/cert.pem",
        "--tlskey",
        "/home/weave/tls/key.pem",
        "-H",
        "tcp://0.0.0.0:12375"
    ],
    "State": {
        "Running": false,
        "Paused": false,
        "Restarting": false,
        "OOMKilled": false,
        "Dead": false,
        "Pid": 0,
        "ExitCode": 1,
        "Error": "",
        "StartedAt": "2015-08-11T21:34:22.304859303Z",
        "FinishedAt": "2015-08-11T21:34:22.692952868Z"
    },
    "Image": "499a6ccbaf6e10396b36ca98cfda6798790117b0b5efc5e8d808654f5805d827",
    "NetworkSettings": {
        "Bridge": "",
        "EndpointID": "",
        "Gateway": "",
        "GlobalIPv6Address": "",
        "GlobalIPv6PrefixLen": 0,
        "HairpinMode": false,
        "IPAddress": "",
        "IPPrefixLen": 0,
        "IPv6Gateway": "",
        "LinkLocalIPv6Address": "",
        "LinkLocalIPv6PrefixLen": 0,
        "MacAddress": "",
        "NetworkID": "",
        "PortMapping": null,
        "Ports": null,
        "SandboxKey": "",
        "SecondaryIPAddresses": null,
        "SecondaryIPv6Addresses": null
    },
    "ResolvConfPath": "/mnt/sda1/var/lib/docker/containers/c269e54807befc16b6e0e53fe280cf6a4ca5f8eb0a0df9e0a69f5af0598fb0c8/resolv.conf",
    "HostnamePath": "/mnt/sda1/var/lib/docker/containers/c269e54807befc16b6e0e53fe280cf6a4ca5f8eb0a0df9e0a69f5af0598fb0c8/hostname",
    "HostsPath": "/mnt/sda1/var/lib/docker/containers/c269e54807befc16b6e0e53fe280cf6a4ca5f8eb0a0df9e0a69f5af0598fb0c8/hosts",
    "LogPath": "/mnt/sda1/var/lib/docker/containers/c269e54807befc16b6e0e53fe280cf6a4ca5f8eb0a0df9e0a69f5af0598fb0c8/c269e54807befc16b6e0e53fe280cf6a4ca5f8eb0a0df9e0a69f5af0598fb0c8-json.log",
    "Name": "/weaveproxy",
    "RestartCount": 0,
    "Driver": "aufs",
    "ExecDriver": "native-0.2",
    "MountLabel": "",
    "ProcessLabel": "",
    "AppArmorProfile": "",
    "ExecIDs": null,
    "HostConfig": {
        "Binds": [
            "/var/lib/boot2docker/tls/ca.pem:/home/weave/tls/ca.pem",
            "/var/lib/boot2docker/tls/server.pem:/home/weave/tls/cert.pem",
            "/var/lib/boot2docker/tls/serverkey.pem:/home/weave/tls/key.pem",
            "/mnt/sda1/var/lib/docker/containers:/mnt/sda1/var/lib/docker/containers",
            "/var/run:/var/run",
            "/proc:/hostproc"
        ],
        "ContainerIDFile": "",
        "LxcConf": [],
        "Memory": 0,
        "MemorySwap": 0,
        "CpuShares": 0,
        "CpuPeriod": 0,
        "CpusetCpus": "",
        "CpusetMems": "",
        "CpuQuota": 0,
        "BlkioWeight": 0,
        "OomKillDisable": false,
        "MemorySwappiness": null,
        "Privileged": true,
        "PortBindings": {},
        "Links": null,
        "PublishAllPorts": false,
        "Dns": null,
        "DnsSearch": null,
        "ExtraHosts": null,
        "VolumesFrom": null,
        "Devices": [],
        "NetworkMode": "host",
        "IpcMode": "",
        "PidMode": "",
        "UTSMode": "",
        "CapAdd": null,
        "CapDrop": null,
        "GroupAdd": null,
        "RestartPolicy": {
            "Name": "",
            "MaximumRetryCount": 0
        },
        "SecurityOpt": null,
        "ReadonlyRootfs": false,
        "Ulimits": null,
        "LogConfig": {
            "Type": "json-file",
            "Config": {}
        },
        "CgroupParent": "",
        "ConsoleSize": [
            0,
            0
        ]
    },
    "GraphDriver": {
        "Name": "aufs",
        "Data": null
    },
    "Mounts": [
        {
            "Source": "/var/lib/boot2docker/tls/ca.pem",
            "Destination": "/home/weave/tls/ca.pem",
            "Mode": "",
            "RW": true
        },
        {
            "Source": "/var/lib/boot2docker/tls/server.pem",
            "Destination": "/home/weave/tls/cert.pem",
            "Mode": "",
            "RW": true
        },
        {
            "Source": "/var/lib/boot2docker/tls/serverkey.pem",
            "Destination": "/home/weave/tls/key.pem",
            "Mode": "",
            "RW": true
        },
        {
            "Source": "/mnt/sda1/var/lib/docker/containers",
            "Destination": "/mnt/sda1/var/lib/docker/containers",
            "Mode": "",
            "RW": true
        },
        {
            "Source": "/var/run",
            "Destination": "/var/run",
            "Mode": "",
            "RW": true
        },
        {
            "Source": "/proc",
            "Destination": "/hostproc",
            "Mode": "",
            "RW": true
        },
        {
            "Name": "82e65d68d3c74e71d89f1dee62d76df474e0f727d82483a1179c75dddeb478b4",
            "Source": "/mnt/sda1/var/lib/docker/volumes/82e65d68d3c74e71d89f1dee62d76df474e0f727d82483a1179c75dddeb478b4/_data",
            "Destination": "/w",
            "Driver": "local",
            "Mode": "",
            "RW": true
        }
    ],
    "Config": {
        "Hostname": "weave-1",
        "Domainname": "",
        "User": "",
        "AttachStdin": false,
        "AttachStdout": false,
        "AttachStderr": false,
        "ExposedPorts": null,
        "PublishService": "",
        "Tty": false,
        "OpenStdin": false,
        "StdinOnce": false,
        "Env": [
            "PROCFS=/hostproc",
            "WEAVE_CIDR=none",
            "DOCKER_BRIDGE=",
            "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
        ],
        "Cmd": [
            "--tls",
            "--tlscacert",
            "/home/weave/tls/ca.pem",
            "--tlscert",
            "/home/weave/tls/cert.pem",
            "--tlskey",
            "/home/weave/tls/key.pem",
            "-H",
            "tcp://0.0.0.0:12375"
        ],
        "Image": "weaveworks/weaveexec:latest",
        "Volumes": {
            "/w": {}
        },
        "VolumeDriver": "",
        "WorkingDir": "/home/weave",
        "Entrypoint": [
            "/home/weave/weaveproxy"
        ],
        "NetworkDisabled": false,
        "MacAddress": "",
        "OnBuild": null,
        "Labels": null
    }
}

**Comment by rade** [comment]
> http://weave.works/guides/weave-and-docker-platform/chapter2/machine-with-weave-proxy.html 

Ah, so we *do* have documentation on how to configure the proxy when using machine on a Mac!

There is something wrong with the volume mounts. What does `docker version` say?

**Comment by abuehrle** [comment]
Misunderstood you. Yes, we are updating them.  

Client:
 Version:      1.8.0-rc3
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   696147b
 Built:        Fri Aug  7 17:36:36 UTC 2015
 OS/Arch:      darwin/amd64

Server:
 Version:      1.8.0-rc3
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   696147b
 Built:        Fri Aug  7 17:36:36 UTC 2015
 OS/Arch:      linux/amd64

**Comment by rade** [comment]
There's the problem. You are running a docker preview release. Which evidently has broken weave! Try to find a way to run 1.7.1.

 @paulbellamy Note the new `Mounts` section in `inspect`. And the absence of a `Volumes` and `VolumesRW` section. Also note that it says `"RW": true` on all the mounts!

**Comment by rade** [comment]
re the machine-with-weave-proxy guide... This should be merged into chapter 1, where we should go straight to a setup with the proxy.

**Comment by abuehrle** [comment]
Yes, that's what i'm doing.  Same with cluster and swam if I can ever get to it ;)

**Comment by rade** [comment]
> So where we want to get to here is the following sequence:
>
>    1. install docker as per docker's instructions, i.e. using the Docker Toolbox
>    2. download weave as per our normal instructions
>    3. perform some extra step, if necessary
>    4. launch weave as per our normal instructions

I think step 3 should become
````
export WEAVEPROXY_ARGS=$(docker-machine ssh weave-1 "cat /proc/\$(pgrep /usr/local/bin/docker)/cmdline | tr '\0' '\n' | grep ^--tls | tr '\n' ' '")
````

which requires some small changes to the weave script.

And it would be great to shorten that further to, say,
````
export WEAVEPROXY_ARGS=$(weave proxy-args docker-machine weave-1)
````
(we can imagine extending this for other backends)

I notice the guide, as it stands, would then *additionally* require you to

````
export DOCKER_CLIENT_ARGS="$(docker-machine config weave-1 | sed 's|:2376|:12375|')"
````
instead of running `eval $(weave env)`.

And it requires running docker as `docker $DOCKER_CLIENT_ARGS ...`.

Both of which is horrific. We must do better. Running docker should not require extra arguments.

**Comment by rade** [comment]
I have raised #1287 for the 1.8.0 breakage, which we need to address for weave 1.0.2.

This issue then is about simplifying the steps required to use weave on a Mac, and documenting them. The latter should be done as part of our main docs, replacing the defunct boot2docker.md, and linked to prominently from the installation section in the README. The guides will need updating too, but that's a separate task.

**Comment by rade** [comment]
@paulbellamy @errordeveloper any thoughts on this? In particular https://github.com/weaveworks/weave/issues/1285#issuecomment-130171411.

**Comment by paulbellamy** [comment]
I'm not sure I follow why you couldn't use `eval $(weave env)` after booting the proxy on docker-machine?

**Comment by rade** [comment]
That's what the guide does. I don't know why.

**Comment by errordeveloper** [comment]
I am not sure, but `weave env` doesn't seem exist...

**Comment by rade** [comment]
`weave env` is on master. 1.0.x has `weave proxy-env`

**Comment by errordeveloper** [comment]
Ok, that makes sense.

**Comment by errordeveloper** [comment]
https://github.com/weaveworks/guides/pull/30#discussion_r37545505
